### PR TITLE
Doron/explicit tool definition option

### PR DESCRIPTION
**Pull request by doron-st**
LMK if it looks good I'll add test and readme.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1 +1,5 @@
-.
\ No newline at end of file
+.
+simppl~=1.0.4
```
Comment (noamBarkai) [inline requirements.txt:2]: I'm not sure I understand this dependency - the tool depends on itself?

File: simppl/cli.py [hunk 2]
```diff
@@ -4,24 +4,31 @@
...
 
 class CommandLineInterface:
 
-    def __init__(self, main_path, module_ascii_logo):
-        self.module_ascii_logo = module_ascii_logo
-        self.module_path = os.path.dirname(main_path)
-        self.module_name = os.path.basename(self.module_path)
-        self.command_line_tools = {}
-        self.tool_name_to_package = {}
+    def __init__(self, main_path: str, module_ascii_logo: str, modules_list: list = None):
```
Comment (noamBarkai) [inline simppl/cli.py:17]: back to static typing? :)

**Review by noamBarkai — COMMENTED** [review]
the general idea looks good. I wonder though if the right approach in terms of usability of the this library is to have the use input a list of tools? and then separately specify the tool she actually wants to run. an alternative might be to have the user set some flag - say ```--skip_tool_scanning``` and then specify the tool of choice by its fully qualified name, including module etc?

File: simppl/cli.py [hunk 3]
```diff
@@ -37,9 +44,9 @@ def print_usage(self):
         return ''
 
     # Add new tools to and return as map<package, tool_name, command_line_tool_run_method>
-    def load_tools(self):
-        script_dir = self.module_path
-        root_dir = os.path.dirname(self.module_path) + '/'
+    def __load_tools(self):
+        script_dir = self.cli_path
+        root_dir = os.path.dirname(self.cli_path) + '/'
```
Comment (zozo123) [inline simppl/cli.py:49]: if possible to use `os.sep`

File: simppl/cli.py [hunk 4]
```diff
@@ -83,15 +89,25 @@ def run(self, argv):
...
+            tool_name = tool.__name__.split('.')[-1]
+            package = tool.__package__
+            self.command_line_tools[package][tool_name] = tool.run
+            self.tool_name_to_package[tool_name] = package
+
     @staticmethod
-    def run_tool(tool_name: str, package: dict, arguments_list):
-        argv = [tool_name] + arguments_list
-        package[tool_name](argv)
+    def __run_tool(tool_name: str, package: dict, argv: List[str]) -> int:
```
Comment (zozo123) [inline simppl/cli.py:105]: it won't break the API? was it always used as a private method?

**Review by zozo123 — COMMENTED** [review]
Looks great! Thanks :)

File: requirements.txt (hunk 1 above)
Comment (doron-st) [reply requirements.txt:2]: hmm. looks like a mistake

File: simppl/cli.py (hunk 2 above)
Comment (doron-st) [reply simppl/cli.py:17]: The more the better, I think

File: simppl/cli.py (hunk 4 above)
Comment (doron-st) [reply simppl/cli.py:105]: yes. I just didn't mark it as private before

**Comment by doron-st** [comment]
> the general idea looks good. I wonder though if the right approach in terms of usability of the this library is to have the use input a list of tools? and then separately specify the tool she actually wants to run. an alternative might be to have the user set some flag - say `--skip_tool_scanning` and then specify the tool of choice by its fully qualified name, including module etc?

@noamBarkai the user doesn't specify a list of tools. the developer does.
The user only specify which tool he wants to use.

File: simppl/cli.py (hunk 3 above)
Comment (doron-st) [reply simppl/cli.py:49]: will do

File: simppl/cli.py [hunk 5]
```diff
@@ -4,24 +4,31 @@
 import argparse
 import sys
 from argparse import ArgumentParser
+from collections import defaultdict
 from functools import wraps
+from typing import List
 
-from colorama import Fore  # , Back, Style
```
Comment (noamBarkai) [inline simppl/cli.py:9]: cleanup?

**Review by noamBarkai — APPROVED** [review]
LGTM

File: README.md [hunk 6]
```diff
@@ -51,13 +51,31 @@ cli enables turning a collection of python executable scripts into a unified cli
...
+### Using cli - developer side:
 - example_module gives an example of how to use CommandLineInterface in your project
 - requirements:
-    - __main__.py - define toolbox logo, constructs and runs the CommandLineInterface. 
+    - __main__.py - define toolbox logo, constructs and runs the CommandLineInterface.
     - __init__.py - set logging configuration
-    - logging_config.ini - python.logging configuration
-    - tools - each script defined as command_line_tool will be automatically added to the toolbox
+
+#### Explicit tool loading mode:
```
[the remark this answers is not in the document]
Comment (doron-st) [reply README.md:60]: Good idea

File: simppl/cli.py (hunk 5 above)
Comment (doron-st) [reply simppl/cli.py:9]: yep

**Pull request merged**